gradio/app.py

```diff
@@ -0,0 +1,318 @@
+STATIC_PATH_LIB = pkg_resources.resource_filename(
+    "gradio", "templates/frontend/static")
+VERSION_FILE = pkg_resources.resource_filename("gradio", "version.txt")
+with open(VERSION_FILE) as version_file:
+    VERSION = version_file.read()
+GRADIO_STATIC_ROOT = "https://gradio.s3-us-west-2.amazonaws.com/{}/static/".format(VERSION)
+
+app = FastAPI()
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+templates = Jinja2Templates(directory=STATIC_TEMPLATE_LIB)
+
+
+###########
+# Auth 
+###########
+
+
+@app.get('/user')
+def get_current_user(request: Request) -> Optional[str]:
+    token = request.cookies.get('access-token')
+    return app.tokens.get(token)
+
+
+@app.get('/login_check')
+def login_check(user: str = Depends(get_current_user)):
+    if app.auth is None or not(user is None):
+        return
+    raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, 
+                        detail="Not authenticated")
+
+
+@app.get('/token')
+def get_token(request: Request) -> Optional[str]:
+    token = request.cookies.get('access-token')
+    return {"token": token, "user": app.tokens.get(token)}
+
+
+@app.post('/login')
+def login(form_data: OAuth2PasswordRequestForm = Depends()):
+    username, password = form_data.username, form_data.password
+    if ((not callable(app.auth) and username in app.auth 
+        and app.auth[username] == password)
+        or (callable(app.auth) and app.auth.__call__(username, password))):
+        token = secrets.token_urlsafe(16)
+        app.tokens[token] = username
+        response = RedirectResponse(url="/", status_code=status.HTTP_302_FOUND)
+        response.set_cookie(key="access-token", value=token, httponly=True)
+        return response
+    else:
+        raise HTTPException(status_code=400, detail="Incorrect credentials.")
+
+
+###############
+# Main Routes
+###############
+
+
+@app.head('/', response_class=HTMLResponse)
+@app.get('/', response_class=HTMLResponse)
+def main(request: Request, user: str = Depends(get_current_user)):
+    if app.auth is None or not(user is None):
+        config = app.interface.config
+    else:
+        config = {"auth_required": True, 
+                  "auth_message": app.interface.auth_message}
+    
+    return templates.TemplateResponse(
+        "frontend/index.html", 
+        {"request": request, "config": config}
+    )
+
+    
+@app.get("/config", dependencies=[Depends(login_check)])
+def get_config():
+    return app.interface.config
+
+
+@app.get("/static/{path:path}")
+def static_resource(path: str):
+    if app.interface.share:
+        return RedirectResponse(GRADIO_STATIC_ROOT + path)
+    else:
+        static_file = safe_join(STATIC_PATH_LIB, path)
+    if static_file is not None:
+        return FileResponse(static_file)
+    raise HTTPException(status_code=404, detail="Static file not found")
```

Comment:
Maybe that FastAPI / Starlette StaticFiles could be used here instead :
https://fastapi.tiangolo.com/tutorial/static-files/

It could also be interesting to see if it can be mixed with the Gzip Middleware :
https://fastapi.tiangolo.com/advanced/middleware/#gzipmiddleware

Reply:
Decided to do this way instead of using `StaticFiles` since we'd like to serve the static files from S3 if `share=True` (it's faster to serve from S3 than to pass through SSH tunnel from a user's computer).

What would `Gzip` be used for here?